File: backend/Models/Quiz.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace prid.Models;

public class Quiz
{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? Description { get; set; }
    public bool IsPublished { get; set; } = false;
    public bool IsClosed { get; set; } = false;
    public bool IsTest { get; set; } = false;
    public DateTimeOffset? StartDate { get; set; }
    public DateTimeOffset? EndDate { get; set; }
    public Database Database { get; set; } = null!;
    public int DatabaseId { get; set; }
    public ICollection<Question> Questions { get; set; } = new HashSet<Question>();
    public ICollection<Attempt> Attempts { get; set; } = new HashSet<Attempt>();
    [NotMapped]
    public string Status { get; set;}= "";
    [NotMapped]
    public string Evaluation { get; set; } = "N/A";
    [NotMapped]
    public int FirstQuestionId { 
        get{
            return Questions.Where(q => q.QuizId == this.Id)
                            .OrderBy(q => q.Order)
                            .Select(q => q.Id)
                            .FirstOrDefault();
        } 
     }

    public string GetStatus(User user)
    {
        string res = Attempts.Any(q => q.StudentId == user.Id) ? "EN_COURS" : "PAS_COMMENCE";
        var StudentAttempts = Attempts.Where(a => a.StudentId == user.Id).OrderBy(a => a.Id).LastOrDefault();
        if(StudentAttempts != null && StudentAttempts.Finish.HasValue){
            res = "FINI";
        }
        return res;
    }

    public string GetTestStatus(User user)
    {
        string res = Attempts.Any(q => q.StudentId == user.Id) ? "FINI" : "PAS_COMMENCE";
        if(EndDate!.Value <= DateTimeOffset.Now){
            res = "CLOTURE";
        }
        return res;
    }

}


public class QuizDTO
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? Description { get; set; }
    public bool IsPublished { get; set; } = false;
    public bool IsClosed { get; set; } = true;
    public bool IsTest { get; set; } = false;

    public DateTimeOffset? StartDate { get; set; }
    public DateTimeOffset? EndDate { get; set; }
    public ICollection<AttemptDto> Attempts { get; set; } = new HashSet<AttemptDto>();
    public string Status { get; set ; } = "";
    public string Evaluation { get; set; } = "N/A";
    public int FirstQuestionId { get; set; }
    public DatabaseDto Database { get; set; } = null!;
}

public class QuizWithIdDto {
    public int Id { get; set; }
}

File: frontend/src/app/models/quiz.ts
import { Type } from 'class-transformer';
import { Database } from './database';
import { Question } from './question';


export class Quiz {
    id?: number;
    name?: string;
    description?: string;
    isPublished?: boolean;
    isClosed?: boolean;
    isTest?: boolean;
    @Type(() => Date)
    startDate?: Date;
    @Type(() => Date)
    endDate?: Date;
    databaseId?: number;
    @Type(() => Database)
    database?: Database;
    status? : string;
    evaluation? : string;
    firstQuestionId? : number;
    questions?: Question[];

    get testStartDate(): string {
        return this.startDate !== null && this.startDate? this.startDate?.toLocaleDateString('fr-BE') : 'N/A';
    }

    get testEndDate(): string {
        return this.endDate !== null && this.endDate? this.endDate?.toLocaleDateString('fr-BE') : 'N/A';
    }

    get iconTooltip(): string {
        if(this.status == "EN_COURS"){
            return "poursuivre";
        }else if(this.status == "PAS_COMMENCE"){
            return "commencer";
        }else if(this.status == "FINI"){
            return "revoir";
        }
        return "";
    }

    get quizType(): string {
        if(this.isTest){
            return "Test";
        }else{
            return "Training";
        }
    }

    get quizStatus(): string {
        if(this.isPublished){
            if(this.isClosed){
                return "CLOTURE";
            }else {
                return "PUBLIE";
            }
        }
        return "PAS_PUBLIE";
    }
}
